Emulated wii input improvements

iwubcode opened the pull request:
Changes made:

* Added 3 various intensities for swing (Slow/Medium/Fast) and for shake (Soft/Medium/Hard)
* Allowed those intensities to be configurable per game
* Make swing more reliable by sending the accelerometer data down each frame (instead of a single frame)
* Implements 'dynamic' mode for swing/shakes which chooses a swing (Slow/Medium/Fast) or shake (Soft/Medium/Hard) based on how long the key is pressed  (matching the button held time to the specific intensity category is configurable)
* Increased the horizontal IR bounds, which allows the cursor to reach the edges in some software like the system-menu
* Adds profile switching for wiimotes (adapted from #6106 ) and also provides game specific input profile switching

**I will not be investigating MotionPlus, so please don't ask!**


Latest windows build:  https://dl.dolphin-emu.org/prs/pr-6222-dolphin-latest-x64.7z

iwubcode commented:
Games I've been testing:

* Red Steel - flipping the table (works with swing command)
* Spectrobes Origins - 'Go' and 'Comeback' actions (works with swing command)
* No More Heroes - Swing left / right / up / down (still not working without adding additional motion when in low stance, completely broken on high stance)

Still to test:

* Spectrobes - move the nunchuck for excavation
* Disaster Day of Crisis - lift the rubble

prog-amateur commented:
Hi, I am having trouble with **The Adventures of Tintin: The Game**.  
There is 2 stages (Unicorn parchments and Globe) where the player has to "swing" pieces of puzzles.  
I have a Wii U pro controller and have setup everything to make it work without success. Maybe I set it up wrongly (if so, please feel free to explain how should I set it up), but I am unable to pass these stages. 

Thank you for your help !

iwubcode commented:
@prog-amateur - thanks for the report.  Unfortunately, I don't have that game but feel free to test these builds to help me know if there's more to do.  Some questions

- Dumb question but I'm assuming you are using the build from this PR?
- What is your input config look like?  What have you tried?
- Do you have "Upward" Wiimote checked? Do you get different results if it is checked?
- Is the swing with the nunchuck or the wiimote?
- Does the game show a picture of how to move or just say something like "swing up"?

This may be similar to a game "No More Heroes", I'm currently investigating it.  I have found that straight "swings" don't work with this game but adding an additional motion will make it work in some cases.  I haven't had a chance to test the game on a real Wii yet so I'm not sure if this is a case where the game actually expects more than just a vertical/horizontal swing or if we've got something wrong with our code.

Thanks again for your feedback!

prog-amateur commented:
Thank you for your reply. 
In my previous post, I have used the term "swing", but after deeper check, I would better use the term "roll". 

To answer your questions : 

**-     Dumb question but I'm assuming you are using the build from this PR?**
There is no dumb question ^^. I am using the latest dolphin build in emulationstation. I don't know exactly which PR it is, but it has not been downloaded from github. Please find the version below : 
![5](https://user-images.githubusercontent.com/33793118/33422166-a0b0a87c-d5ac-11e7-9860-8266d50fe88f.png)

**-     What is your input config look like? What have you tried?** 
My default input was this (but I have tried many things) : 
![1](https://user-images.githubusercontent.com/33793118/33421871-720abf40-d5ab-11e7-971f-7c28adc8e8ff.png)
![2](https://user-images.githubusercontent.com/33793118/33421872-7223f4e2-d5ab-11e7-867c-ecead1097365.png)
![3](https://user-images.githubusercontent.com/33793118/33421873-72414f74-d5ab-11e7-9f02-913346f490f0.png)

**-     Do you have "Upward" Wiimote checked? Do you get different results if it is checked or unchecked?**
"Upward" was unchecked by default, but I have tried with the checked option too
**-     Is the swing with the nunchuck or the wiimote?**
On the screen intructions, the roll is shown with an upward wiimote, so I assume it is with a wiimote. 
**-     Does the game show a picture of how to move or just say something like "swing up"?**
Same as above, the instruction shows an upward wiimote rolling clockwise/counterclockwise. I do have mapped the wiimote swing inputs with my controller stick, but there is no change.

prog-amateur commented:
Hi, I finally succeed to map the roll input like this : 
![4](https://user-images.githubusercontent.com/33793118/33429529-1a5c4c72-d5c5-11e7-90f3-268249e1c47a.png)
2 cases : 

- If you have Upward unchecked (my case), then input the Forward/Backward of the Tilt section as above. 
- If you have the Upward checked, then you have to input the Left/Right of the Tilt section. 

Awesome !

HernanRC2 commented:
**- Disaster Day of Crisis - lift the rubble -------->** 
This works in very old releases, like SVN 5928, but not in the current one. I don't know if this can be used to fix in the current version.

HernanRC2 commented:
@iwubcode I'm sorry, but I don't know how to use this code, I'm new here. Let me find out how to do it.

iwubcode commented:
@HernanRC2  - click "show all checks" -> "Details" (for the build type, ex: windows) -> that'll open the buildbot window, then choose "build properties" tab -> copy "build url".

Here's the windows build:  https://dl.dolphin-emu.org/prs/pr-6222-dolphin-latest-x64.7z

HernanRC2 commented:
@iwubcode I think it does not work. I configured in this way, pressing the "O" and "P" keys at the same time:
![1](https://user-images.githubusercontent.com/34307628/33767389-1df00e50-dc00-11e7-91a3-fda2b49aa9a8.jpg)
![2](https://user-images.githubusercontent.com/34307628/33767402-2597055a-dc00-11e7-9280-8e22da540df1.jpg)

iwubcode commented:
@HernanRC2 - thanks.  One last thing to try would be to assign a down swing as well.  Then go from down to up to represent an upward motion.

If that doesn't work, I will look at it eventually.  I own the japanese copy and didn't realize the text would be as important as it is.  So, I'm now waiting for a PAL copy to arrive before progressing further.

kurayami6 commented:
Could you try Escape from Bug Island. It has some really awkward first person aiming.

HernanRC2 commented:
> @HernanRC2 - thanks. One last thing to try would be to assign a down swing as well. Then go from down to up to represent an upward motion.

@iwubcode This does not work either, but thanks anyway. I will follow your progress closely.

iwubcode commented:
Between the holidays and some personal issues, I haven't had much of a chance to look at this.  Sorry all!

Finally got to the point where the fossil needs to be excavated in Spectrobes.  Can confirm the nunchuck shake in [issue 9611](https://bugs.dolphin-emu.org/issues/9611) seems to be fixed.  It is working.

The wiimote shake is actually more problematic as I needed both a Swing+Shake to finish the hammering section (swing wiimote vertically).  I am going to investigate why this is the case but I'm guessing it is due to improper intensities for swing.

I did receive my Disaster copy but have not had time to play it to get to the rubble section.

iwubcode commented:
I also recently got a copy of Final Fantasy Fables: Chocobo's Dungeon.  It's a cute game!  Only played a few minutes of it.  I did not run into issues like the thread mentioned (maybe because it was posted in 2012) but I did notice sorting (swinging the Wiimote) was a bit sporadic on the master branch, it didn't work the first time.  For this PR, I had no issue.

Now to get back to testing Disaster!

JMC47 commented:
Could you make the default "range" of IR (not technically motion, but close enough.) wider for games that expect you to be pointing closer to the edges?  If you try the system menu with the default values, you can't reach any of the corners or edges.

iwubcode commented:
@JMC47 - yeah, I've experienced that myself.  I'll add that as another item to do.  Progress is slow but happening :)

iwubcode commented:
@JMC47 - keep meaning to talk to you on IRC but things have been really busy lately.  Have you or can you test "Baman Brave and the Bold" to see if this PR fixes the issue?  One user reported it didn't and I just wanted to get confirmation.  I also need to discuss some design ideas but I'll have to do that over chat!

For everyone else, sorry progress has been non-existent lately.  I did get to the problematic area in Disaster and can confirm the issue but haven't figured out the problem yet.

iwubcode commented:
Finally a bit of time to work on this.  I implemented wiimote swing/shake intensities that are configurable by game.

Currently the controls are only configurable via the configuration (no UI yet).  Example of control configuration in a profile:

```
[Wiimote]
Shake/Y = A
ShakeSoft/Y = B
ShakeHard/Y = C

Swing/Forward = E
SwingFast/Forward = F
SwingSlow/Forward = G
```

I'll add an example on how to configure per game later (I'm actually not sure what the full setting path is myself because of the new config system).  I just put some random defaults for Fast/Slow and Soft/Hard that seemed to fit Boom Blox.

I'll add Nunchuck support next.


For Boom Blox (and likely any game similar), the "Shake" seems to royally mess up the cursor.  I'll likely investigate this in the future as it's really annoying..

riking commented on Source/Core/Core/HW/WiimoteEmu/Attachment/Nunchuk.cpp:
```diff
@@ -40,6 +41,8 @@ Nunchuk::Nunchuk(ExtensionReg& reg) : Attachment(_trans("Nunchuk"), reg)
 
   // swing
   groups.emplace_back(m_swing = new ControllerEmu::Force(_trans("Swing")));
+  groups.emplace_back(m_swingSlow = new ControllerEmu::Force(_trans("SwingSlow")));
```
these.... _are_ supposed to be translated, right...?

Line 35, "C" and "Z" aren't being translated but these are.

JosJuice replied:
C and Z are intentionally not translatable because there isn't anything reasonable to translate them to other than C and Z.

riking replied:
I guess really, I'm confused over whether this string is for UI display or for loading from config files.

edit: looking at it again, that's what the ConfigInfo structs are for. OK

riking commented on Source/Core/Core/HW/WiimoteEmu/Attachment/Nunchuk.cpp:
```diff
@@ -90,9 +109,15 @@ void Nunchuk::GetState(u8* const data)
 
   // swing
-  EmulateSwing(&accel, m_swing);
+  EmulateSwing(&accel, m_swing, Config::Get(Config::NUNCHUCK_INPUT_SWING_INTENSITY_MEDIUM), false, true);
```
When are we getting "real accelerometer" mode :P

iwubcode commented:
HdkR helped me figure out how to set the settings in the game ini.

For instance, you can do this:

```
[Wiimote.Swing]
Medium = 3.2

[Wiimote.Nunchuck_Swing]
Medium = 4.3
```

This should help fix games that need a harder or softer shake/swing for wiimotes or nunchuk (I know I need to fix the name).

RunicSSB commented:
Can you test New Play Control! Donkey Kong: Jungle Beat? The parts where you have to shake the wiimote and nunchuk rapidly aren't working (it worked fine in 4.0.2, though).

iwubcode commented:
@RunicSSB - I do not own that game and have no plans to buy it but if someone else wants to give it a whirl and let me know if it works that'd be great.

The windows build link is at the top and the instructions for changing the strength for each preset (and how to bind keys to them) is in my most recent comments.

It's possible it's still broken, I haven't fixed all the games I own yet.

RunicSSB commented:
I see your comments, it says what I'm supposed to do, but where do I do it?

RunicSSB commented:
It's supposed to go in WiimoteNew.ini, right? I tried doing all that stuff, and it helped a bit, but not much. Now I can have both the wiimote and nunchuk on one button and activate just by holding it, but it seems random whether or not it'll work and it only does several at a time (if any), whereas before it did at least 20.

iwubcode commented:
@RunicSSB -

Things like the input key settings go in your profile for instance (you can create a profile via the UI but just need to add `ShakeSoft`, `SwingSlow` etc manually):

```
[Wiimote]
Shake/Y = A
ShakeSoft/Y = B
ShakeHard/Y = C

Swing/Forward = E
SwingFast/Forward = F
SwingSlow/Forward = G
```

If you want to change the intensities, they go in the game ini, example (random numbers):

```
[Wiimote.Swing]
Slow = 3.8
Medium = 4
Fast = 4.4


[Wiimote.Shake]
Soft = 1.1
Medium = 2.3
Hard = 3.3

[Wiimote.Nunchuck_Swing]
Slow = 2.2
Medium = 3.3
Fast = 4.8
```

Are you trying to do 'shake' or 'swing'?  What kind of configuration options did you change?

I can't recall what the old default might have been on 4.0.2 for swing/shake (or if Wiimote/nunchuk was different), not sure if @JMC47 remembers?

RunicSSB commented:
Where's the folder location for the "game ini"?

iwubcode commented:
@RunicSSB - right-click on game then choose "Properties", then look at the bottom left for "Edit User Config", click that and add your settings.

RunicSSB commented:
I edited my profile to change all of my "Shake" binds to "ShakeHard". That did nothing. I went to the game's ini and set "ShakeSard" to 10. Nothing. Then I did the exact opposite, set it to 1. Still nothing. I don't know what I'm doing wrong.

RunicSSB commented:
I tried raising and lowering the shake strength in the game's .ini in a bunch of different ways, but nothing changed.

iwubcode commented:
@HernanRC2 - I can confirm that this PR fixes the rubble issue in *Disaster Day of Crisis* (in fact it might work in master as well, I haven't tested it yet).

The fix is the same as No More Heroes slashes.  You need to set a forward (or possibly backward) swing in addition to the Up/Down.

So what I did was hold the key for forward swing, then switch back and forth between Up/Down swings.  The rubble on the old lady was no more!

lioncash commented on Source/Core/Core/HW/WiimoteEmu/WiimoteEmu.cpp:
```diff
@@ -232,9 +232,9 @@ void Wiimote::Reset()
-  memset(m_shake_step, 0, sizeof(m_shake_step));
-  memset(m_shakeSoft_step, 0, sizeof(m_shakeSoft_step));
-  memset(m_shakeHard_step, 0, sizeof(m_shakeHard_step));
+  memset(m_shake_step.data(), 0, sizeof(m_shake_step));
```
You can just do `m_shake_step = {};` or `m_shake_step.fill(0);` instead of using memset now.

lioncash commented on Source/Core/Core/HW/WiimoteEmu/WiimoteEmu.h:
```diff
@@ -247,9 +248,13 @@ class Wiimote : public ControllerEmu::EmulatedController
   ControllerEmu::Buttons* m_buttons;
   ControllerEmu::Buttons* m_dpad;
   ControllerEmu::Buttons* m_shake;
+  ControllerEmu::Buttons* m_shakeSoft;
```
`m_shake_soft`, ditto for other camelCased names added

JMC47 commented:
I finally got a chance to test this - looks good so far.  There's some testing I didn't do tonight 'cause I'm lazy (NSMBWii's tilting platforms not having the full range in master...) but in general this works way better.

iwubcode commented:
I added an initial implementation of the dynamic swing/shake.  It's still not perfect and there are a few hardcoded values that I might make be tweakable by game but it does work.  I'm able to throw light/power throws in Boom Blox with basically a single button!  (Boom Blox like other games that require swings, needs a forward motion in addition to the downard motion for throwing a ball)

I also updated the IR range to fix games that previously couldn't reach the edges (ex: the system menu).  This was requested by @JMC47 .  Let me  know if there's any regressions here.

JMC47 commented:
Functionally looks good to me.

leoetlino commented:
This needs a rebase and a small lint fix.

iwubcode commented:
This next change adds profile 'cycling' (gamecube coming in a future PR) through input profiles either forward or backward.  Which is adapted with some code changes from Barath-Kannan's PR #6106 .  Other than code changes, it now supports Qt and WX UIs.

It then extends this capability to add game-configurable profile cycling.  You can do it like so:

```
[Wiimote.InputProfiles]
List = "BoomBlox.ini,BoomBlox2.ini"
```

I still need to test some edge cases but overall the effect works and should be useful not only for games that have vastly different gameplay experiences (think Disaster Day of Crisis) but also can help cut down on the amount of button inputs by having allowing the power-user to do context-switching based on the current gameplay elements.

JMC47 commented:
I can report that Tales of Symphonia works fine here and in master.

iwubcode commented:
Thanks @JMC47 for all the testing!!!

If you have settings that work well for Jet Rocket, let me know.  We could then set it in the game ini.

lioncash commented on Source/Core/InputCommon/InputProfile.cpp:
```diff
@@ -0,0 +1,187 @@
+    const std::string device_profile_root_location(File::GetUserPath(D_CONFIG_IDX) + "Profiles/" + device_configuration->GetProfileName());
+
+    std::vector<std::string> result;
+    result.resize(profiles.size());
```
`std::vector<std::string> result(profiles.size());`

lioncash commented on Source/Core/InputCommon/InputProfile.cpp:
```diff
@@ -0,0 +1,187 @@
+    }
+
+    std::vector<std::string> result;
+    std::set_intersection(profiles.begin(), profiles.end(), profiles_from_setting.begin(), profiles_from_setting.end(), back_inserter(result));
```
Missing `std::` on back_inserter

lioncash commented on Source/Core/InputCommon/InputProfile.cpp:
```diff
@@ -0,0 +1,187 @@
+    return result;
+  }
+
+  std::vector<std::string> GetMatchingProfilesFromSetting(const std::string& setting, const std::vector<std::string> profiles, InputConfig* device_configuration)
```
`profiles` was likely meant to be taken by reference.

lioncash commented on Source/Core/InputCommon/InputProfile.cpp:
```diff
@@ -0,0 +1,187 @@
+    Backward = -1
+  };
+
+  int s_wiimote_profile_index = 0;
```
Please don't make this file-static-based. All of this behavior should probably be put in a class, rather than a flat namespace, and be driven by the front-end in some form.

iwubcode replied:
I did this mostly because I'm trying to A) keep everything encompassed in one place.  B) Was trying to avoid exposing too much to the UI.

The original implementation of this logic had the index in the hotkey controller interface which I didn't agree with.

Where do things typically get stored?  Even if this was a class, I'm not sure how I feel about the UI 'owning' it directly.

lioncash replied:
If hotkeys are driven by the UI, I really don't see how controller profiles are an exception to this, given there needs to be interaction from the UI to manipulate profiles, therefore the UI owning it would make the most sense here. Realistically may do what you want in terms of where it actually goes as long as it's not static-based like this. This adds "yet another thing to undo/fix" in terms of working towards libdolphin otherwise.

iwubcode replied:
Yeah hotkeys are driven by the UI but I don't really think they should be.

Ideally, I feel like everything should be separated out.  Hotkey ownership would be managed by each type (ex: graphics hotkeys owned by a graphics class, inputs hotkeys owned by a input class, etc).  The UI would have specific touch points into these classes to get at the data it needs or to call the actions it needs to run.  Each class would have its own responsibility and ownership.  UI would never know about specific hotkey interactions, it would just say `handle_hotkeys()`  If you know about the MVVM paradigm, I feel like we're breaking it at the moment.

All that being said, if putting it in a class that is then owned by the UI is the quickest way to get this solved, then I'll likely go for that.

And `libdolphin`?  That's the first I've heard of it.  What would that be used for?

lioncash replied:
It's the quickest way to get it approved (by me, if that matters).

also libdolphin would allow the core to be use separately from the project (i.e. for some other application purpose; think something along the lines of [Unicorn](https://www.unicorn-engine.org/)

iwubcode replied:
Getting the code approved by you is always a goal for me because I know you strive for high-quality code and that's something I aim for as well.  Thank you for taking the time to review.  I will take a look at this in the next day or two.

Also, appreciate the insight into `libdolphin`, very interesting!

lioncash commented on Source/Core/InputCommon/InputProfile.cpp:
```diff
@@ -0,0 +1,187 @@
+    return profiles[profile_index];
+  }
+
+  void UpdateProfile(const std::string& profile_filename, const std::vector<ControllerEmu::EmulatedController*> controllers)
```
`controllers` was likely meant to be taken by reference.

lioncash commented on Source/Core/InputCommon/InputProfile.cpp:
```diff
@@ -0,0 +1,187 @@
+  {
+    std::size_t size = device_configuration->GetControllerCount();
+
+    std::vector<ControllerEmu::EmulatedController*> result;
```
This can just be:

```cpp
std::vector<ControllerEmu::EmulatedController*> result(size);
```
Given it's only a vector of pointers and default initializing those are trivial.

lioncash commented on Source/Core/InputCommon/InputProfile.cpp:
```diff
@@ -0,0 +1,187 @@
+
+namespace InputProfile
+{
+  void NextWiimoteProfile()
```
There seems like quite an awful lot of duplication going on here for what is essentially just a simple change in direction. Perhaps consider factoring this out in such a way where it's minimized more. (e.g. an implemetation function that is just called from both prev and next variants, which only serve to provide the direction as an argument to that implementation function).

iwubcode replied:
Yeah, it did come to mind but I wanted to get all my requirements met before going back and fleshing out the code design.

I'll add this to my to-do list.  Thanks @lioncash !

iwubcode commented:
Most recent commit added configurable swing/shake for the wiimote at each intensity to the Qt UI only.  I ended up using a tab view since it was easiest.

@spycrab @JMC47 @MayImilae - interested in your opinion on whether this is the best solution.

I had originally thought of a listbox but since the capabilities aren't dynamic (as I had originally hoped for), the tab solution made more sense.  Not sure if there is any other option.

I also contemplated adding a info box (at the top of the controls) - this would do two things:

* Tell users that swing is often a combination of two motions (ex: up swing for No More Heroes/Disaster Day of Crisis/etc is a Up + forward motion).
* Explain what the "dynamic" option is

Once the general UI is signed off on, I'll add the same UI layout for Nunchuk.

iwubcode commented:
Thinking about it more, only games that are motion heavy need the various intensity options (Boom Blox, Wii Sports, etc).  I could see the argument that adding all this extra configuration for only a select number of titles adds complexity where it isn't needed.

In that case, I'm fine keeping those options as features that can only be set by editing the profile's ini file.

iwubcode commented:
Latest change will now allow for cycling on a per controller basis.  This way supports multiplayer and will line-up with some other changes that are being made.

New per-game setting is like this:

```
[Wiimote.1]
InputProfiles = "BoomBlox.ini,BoomBlox2.ini"
```

JosJuice commented on Source/Core/Common/StringUtil.cpp:
```diff
@@ -369,6 +369,18 @@ std::string JoinStrings(const std::vector<std::string>& strings, const std::stri
   return joined.substr(0, joined.length() - delimiter.length());
 }
 
+std::string Trim(const std::string& str)
```
This seems very similar to the StripSpaces function that already exists in this file. Would it be possible to use that instead?

iwubcode replied:
Hmm, I was surprised when I thought a trim function didn't exist.  Anyway, the one issue I have with `StripSpaces` is it returns empty string if there aren't any spaces.  That probably won't work for me as it's user defined (so it might not have a space in it).  The function is used all over the place, not sure I should change the functionality?

iwubcode replied:
Also, not a huge fan with the name (it implies it removes all spaces) but if it works, I'm happy to adopt it.

iwubcode replied:
I suppose I can use `StripSpaces` then check if it's empty and return the actual string

JosJuice replied:
> Anyway, the one issue I have with `StripSpaces` is it returns empty string if there aren't any spaces.

That's not what it does as far as I know. When reading the code, keep in mind that it's using `find_first_not_of`, not `find_first_of`.

iwubcode commented:
I still need to get to a couple of @lioncash 's comments but anyone is free to review when they have time.  Functionally, I think it's ready.

lioncash commented on Source/Core/InputCommon/InputProfile.h:
```diff
@@ -0,0 +1,45 @@
+    std::string GetWiimoteInputProfilesForGame(int controller_index);
+
+    int m_wiimote_profile_index = 0;
+    const int m_display_message_ms = 3000;
```
This can be a constexpr variable in the cpp file, given there's no dependency on it in the header, nor a need for it to be instance-based (since it's a constant value).

JMC47 commented:
Is this ready to be reviewed, merged?  I'd love to have this in the progress report, especially with the Hybrid Wii Remote removal.

JosJuice commented on Source/Core/Core/HW/WiimoteEmu/WiimoteEmu.cpp:
```diff
@@ -271,6 +374,9 @@ Wiimote::Wiimote(const unsigned int index) : m_index(index), ir_sin(0), ir_cos(1
   // swing
   groups.emplace_back(m_swing = new ControllerEmu::Force(_trans("Swing")));
+  groups.emplace_back(m_swing_slow = new ControllerEmu::Force(_trans("SwingSlow")));
+  groups.emplace_back(m_swing_fast = new ControllerEmu::Force(_trans("SwingFast")));
```
CamelCase shouldn't be used in user-facing strings in this way.

JosJuice replied:
Oh, okay. In that case, you should remove the `_trans` macro instead, since it has no effect if the string isn't going to be displayed in the GUI.

JosJuice commented on Source/Core/Core/HW/WiimoteEmu/WiimoteEmu.cpp:
```diff
@@ -284,6 +390,39 @@ Wiimote::Wiimote(const unsigned int index) : m_index(index), ir_sin(0), ir_cos(1
   m_shake->controls.emplace_back(new ControllerEmu::Input(ControllerEmu::Translate, _trans("Z")));
 
+  groups.emplace_back(m_shake_soft = new ControllerEmu::Buttons(_trans("ShakeSoft")));
+  // i18n: Refers to a 3D axis (used when mapping motion controls)
```
The `i18n` comment must be on the line immediately before the `_trans`. Right now it's two lines above, which won't work.

JosJuice replied:
> Are you saying that the comment is read by some process?

Yes, xgettext.

> I can refactor..

That's not necessary. You should be able to just move the comments one line down (putting the comment in the middle of the statement).

iwubcode replied:
Thanks for the info!  Just to be sure, if these aren't used in the UI and we remove the `_trans` macro, do I still need to update the comment?

JosJuice replied:
If you remove the `_trans` macro, you should remove the corresponding `i18n` comment too.

JosJuice commented on Source/Core/Core/HotkeyManager.cpp:
```diff
@@ -69,6 +69,23 @@ const std::string hotkey_labels[] = {
+    _trans("Next Profile for Wiimote 4"),
+    _trans("Previous Profile for Wiimote 4"),
+    _trans("Next Game Profile for Wiimote 4"),
+    _trans("Previous Game Profile for Wiimote 4"),
```
We always use `Wii Remote` in user-facing strings, not `Wiimote`.

iwubcode commented:
@JMC47 - functionally this is ready.  Assuming there's not many other changes, this should be ready to merge.

I was hoping to get a few other ini changes in here.  See my comment about Tides of Destiny and Jett Rocket.  I also thought of making an ini for Wii Sports for the soft/medium/hard swings but haven't gotten to it and figure those changes could be in a follow up PR.

JMC47 commented:
I wouldn't worry about INI stuff until afterwards.  It's just such a huge improvement and Quality of Life thing that I'd rather see it now.  Lose a feature with Hybrid Wii Remotes, but get much better Emulated Wii Remotes.  I can pair them up in the progress report nicely, and this has been around a while.

JMC47 commented:
I actually thought of something - Does this have a remedy for games that htink the Wii Remote is too close to the screen?  our IR distance is wrong for those games I think?

iwubcode commented:
@JMC47 - if you're talking about the horizontal limitation, yes, it should fix that.  If you're talking about something else, no, it only has the changes in the first post.  What games are affected by that issue?  What are the symptoms of the problem?

iwubcode commented:
After re-testing, it turns out that Red Steel / Spectrobes didn't need that 'bit regression' or the change to the nunchuk up direction.   Just like all other games, the swing of the nunchuk needs to apply some forward (or likely backward) motion in addition to the other motion.

**Spectrobes:**

Go - Forward + Down Wiimote Swing (or up)
Come-Back - Forward + left wiimote swing (or right)

Hammer - Shake Y + Down / Up Swing seemed to be the most effective, it also worked with up/down swings but might need to increase intensity for more consistency
Music-notes - Forward + Down Wiimote Swing for purple notes, blue notes needed forward + down swing for nunchuk

**Red Steel:**

Table - needs forward + down swing of nunchuk

Techjar commented on Source/Core/InputCommon/InputCommon.vcxproj.filters:
```diff
@@ -202,8 +203,9 @@
     <Text Include="CMakeLists.txt" />
   </ItemGroup>
-</Project>
+</Project>
```
Why remove the newline?

iwubcode replied:
Hmm, I don't believe it was intentional, I can revert this change.

Techjar commented on Source/Core/InputCommon/InputProfile.cpp:
```diff
@@ -0,0 +1,221 @@
+    control_section->Get("WiimoteProfile3", &result);
+  case 3:
+    control_section->Get("WiimoteProfile4", &result);
+  };
```
You can use `StringFromFormat` here instead of a switch statement.

JMC47 commented:
@iwubcode I'm going to try the latest version and play around with getting values for the games I mentioned.

JMC47 commented:
@iwubcode do you know of a good in-game example to show off Dynamic Shake/Swing strength?  Is it enabled by default?

iwubcode commented:
@JMC47 - we have defaults set up for Boom Blox but the numbers likely need more tweaking.  Any games that needs a variety of motion strengths would likely be a good candidate.

It's enabled by default, you just have to modify the controller profile to set up the keys for dynamic swing/shakes.  It only works for the wiimote but all the variables are configurable.  In the next day or two I'll summarize the full ini/controller information.

HernanRC2 commented:
**Disaster Day of Crisis - lift the rubble**
I can confirm that the problem is fixed (master also) and work with a single up+forward swing (Wiimote+Nunchuk at the same time).
Thanks.

DjBlades commented:
@JMC47 Let's Tap would be completely impossible to play without this, built entirely around different shake strengths. Could be a good example.

iwubcode commented:
Here's all the configurable options in this PR...  ( @JMC47  )

**Multi Profile Configuration**

Unlike before, games can now support multiple configuration profiles.  Instead of specifying a single entry, you can now specify multiple in a comma-separated list or specify a directory which will be searched recursively.

Multiple configuration profiles can be swapped via the hotkey manager on a per-game basis or based on all the controller profiles in your profile directory.  The first profile in the game ini list will be used by default when launching the game.

Example game ini configuration:
```
[Controls]
WiimoteProfile1="BoomBlox,BoomBloxAnother"
WiimoteProfile2="BoomBlox2, BoomBlox2Other"
```

**Game Specific Wii motion Intensities**

Games now can support 3 swing/shake intensities and are configurable on a game basis.   Examples below.

Configuring wiimote swing intensities:
```
[Wiimote.Swing]
Slow = 3.8
Medium = 4
Fast = 4.4
```

Configuring wiimote shake intensities:
```
[Wiimote.Shake]
Soft = 1.1
Medium = 2.3
Hard = 3.3
```

Configuring nunchuk swing intensities:
```
[Wiimote.Nunchuck_Swing]
Slow = 2.2
Medium = 3.3
Fast = 4.8
```

Configuring nunchuk shake intensities:
```
[Wiimote.Nunchuck_Shake]
Soft = 2.2
Medium = 3.3
Hard = 4.8
```

**Dynamic Motions**

Wiimote Swings/Shakes can be defined as "dynamic".  This means that based on how long the key is held down, determines whether the result is a 'slow', 'medium', or 'fast' swing (or a 'soft', 'medium', or 'hard' shake).

Here is an example for the controller profile ini for swings:
```
Swing Dynamic/Up = UP
Swing Dynamic/Down = DOWN
Swing Dynamic/Left = LEFT
Swing Dynamic/Right = RIGHT
```

Additionally, there are many configurable options for dynamic motions that are configurable per-game.

Examples for configuring swings  (the settings listed are the defaults):
```
[Wiimote.Dynamic_Swing]
FramesHeldFast=100
FramesHeldSlow=30
FrameCount=30
```
Here *FramesHeldFast* indicates _at least how many_ frames must be held for a fast swing.  Whereas *FramesHeldSlow* indicates _the maximum_ frames allowed to constitute a slow swing.  A 'medium' swing is anything in between those two numbers.  *FrameCount* indicates how long the swing should take affect for the game to register it properly.

JMC47 commented:
Take this with a grain of salt, but, an android user claims this breaks the shield attack in Twilight Princess with Nunchuk Shake Y

https://forums.dolphin-emu.org/Thread-strange-emulated-wiimote-behaviour-from-dolphin-50-8443-onwards

iwubcode commented:
Thanks @JMC47 , I will look into it.  I have that game but haven't played it on Dolphin so I will need to get to that point.

iwubcode commented:
@JMC47  - I pulled the latest build and tested the Twilight Princess game with a 100% save.  I could not reproduce the issue the user was having.  Shake 'y' worked fine for the Nunchuk and even swing (forward + up) worked on the newer build! (testing on an older build, swing of course didn't work!)

DjBlades commented:
How exactly do you specify directories for the multiple profile switching? Specifying multiple regular profiles wasn't switching correctly for me, but Dolphin can't find the directory either. I've tried both in quotes and without.
Current line is: 
WiimoteProfile1 ="D:\Launchbox\Emulators\Dolphin\User\Config\Profiles\Wiimote\Galaxy"

iwubcode commented:
@DjBlades  - you should be able to do `WiimoteProfile1 = Galaxy` for directory, or you could specify `WiimoteProfile1 = Galaxy/Profile1, Galaxy/Profile2`.  Or you can mix and match directories or profiles.  Keep in mind if you specify a profile, it is expected to _not_ have the .ini on the end (I think that's a bit weird but that's how it worked before).  I didn't test absolute pathes, that might be something that needs to be fixed.  All the pathes are relative to the `Profiles\Wiimote` directory.

Please let me know if you have questions, comments, or issues.

iwubcode commented:
@DjBlades - it should pick whatever ini is first.  If you had two directories, it'll pick the first one in the first directory.  If you have a file before a directory, it should pick that first.  First for a directory _should_ be alphabetical but maybe the function I used doesn't preserve order.

It's possible we need to do some sorting of the list of entries.

What does your setting look like?

DjBlades commented:
GameIni is

DjBlades commented:
GameIni is:
[Controls]
WiimoteProfile1 = Galaxy

Galaxy Folder just has three profiles, Galaxy, GManta and GRolling.

iwubcode commented:
@DjBlades - my guess is it is putting capitalization _before_ lower-cased fields.  So 'Galaxy' would be at the end.

Not sure what OS you're on but I tested this on Windows by creating three text files with your name and using 'ls'.  Galaxy was at the bottom.  This is different than what I expect where lowercase should be at the top (as it is when viewing the files on the filesystem).  Thanks for reporting!

For now, I'd suggest just renaming.  Either:  '1Galaxy', 'GMain', or something similar.

DjBlades commented:
1Galaxy works fine. I'll use that then, thanks for the suggestion!
Also, I noticed in the progress report you're putting default strengths in Game INIs. I've got some strengths for Let's Tap if you want to make use of them:
[Wiimote.Shake]
Soft = 0.1
Medium = 0.5
Hard = 0.9
Works alright on default settings, but it's better if you set the sensitivity to +1 in the game's options menu.

And again, thanks for putting the work into such a useful feature!
